### PR TITLE
Update QAT recipe to match full_finetune_distributed

### PR DESCRIPTION
**Pull request by andrewor14**
**Summary:** Many changes have gone into the full_finetune_distributed recipe but were not reflected on the equivalent qat_distributed recipe. This commit brings the latter up to date, adding features like FSDP2, configurable FSDP sharding, and profiling.

**Test Plan:**
Ran QAT and full fine-tune on 6 A100 GPUs using the default Llama3-8B configs on the alpaca dataset:
```
tune run --nnodes 1 --nproc_per_node 6 qat_distributed --config llama3/8B_qat_full
tune run --nnodes 1 --nproc_per_node 6 full_finetune_distributed --config llama3/8B_full
```

Microbenchmarks:
```
QAT average throughput per GPU: 205.789 tok/s
Full fine-tune average throughput per GPU: 221.237 tok/s

QAT average loss in first and last 100 steps: 1.08369 -> 0.554201
Full fine-tune average loss in first and last 100 steps: 1.06767 -> 0.534989

QAT max memory reserved on rank zero: 17.166 GB
Full fine-tune max memory reserved on rank zero: 17.041 GB
```

QAT evaluation results (quantized):
```
|  Tasks  |Version|Filter|n-shot|    Metric     | Value |   |Stderr|
|---------|------:|------|-----:|---------------|------:|---|------|
|wikitext |      2|none  |     0|word_perplexity|14.2488|±  |N/A   |
|         |       |none  |     0|byte_perplexity| 1.6435|±  |N/A   |
|         |       |none  |     0|bits_per_byte  | 0.7167|±  |N/A   |
|hellaswag|      1|none  |     0|acc            | 0.5666|±  |0.0049|
|         |       |none  |     0|acc_norm       | 0.7490|±  |0.0043|
```

Full fine-tune evaluation results (quantized):
```
|  Tasks  |Version|Filter|n-shot|    Metric     | Value |   |Stderr|
|---------|------:|------|-----:|---------------|------:|---|------|
|wikitext |      2|none  |     0|word_perplexity|28.3989|±  |N/A   |
|         |       |none  |     0|byte_perplexity| 1.8697|±  |N/A   |
|         |       |none  |     0|bits_per_byte  | 0.9028|±  |N/A   |
|hellaswag|      1|none  |     0|acc            | 0.5406|±  |0.0050|
|         |       |none  |     0|acc_norm       | 0.7190|±  |0.0045|
```

**Reviewers:** ebsmothers, kartikayk, RdoubleA

**Subscribers:** ebsmothers, kartikayk, RdoubleA

### CODE REVIEW

**Review by kartikayk — COMMENTED** [review]
@andrewor14 since this is mostly a direct update, can you add some screenshots which show memory parity and parity around loss and tokens/sec for a few steps? I can accept once we know there isn;t anything off and things are on par between FSDP1 and FSDP2

**Review by kartikayk — APPROVED** [review]
Thanks so much for the detailed testing and benchmarking. Really appreciate the thoroughness on this!

File: recipes/qat_distributed.py [hunk 1]
```diff
@@ -137,6 +133,9 @@ def __init__(self, cfg: DictConfig) -> None:
         # Training cfg
         self._resume_from_checkpoint = cfg.resume_from_checkpoint
         self._gradient_accumulation_steps = cfg.gradient_accumulation_steps
+        self._fsdp_sharding_strategy = torch.distributed.fsdp.ShardingStrategy[
+            cfg.get("fsdp_sharding_strategy", "FULL_SHARD")
+        ]
```
Comment (ebsmothers) [inline recipes/qat_distributed.py:138]: ah crap we don't need this (sorry I should've deleted from the full finetune recipe too). This is supplanted by the fsdp_reshard_after_forward config

**Comment by ebsmothers** [comment]
Looks great! Unrelated but didn't I remember that QAT used a fair bit more memory than vanilla full finetune? Based on the benchmark results, is that no longer the case?

**Review by ebsmothers — APPROVED** [review]

**Comment by andrewor14** [comment]
> Looks great! Unrelated but didn't I remember that QAT used a fair bit more memory than vanilla full finetune? Based on the benchmark results, is that no longer the case?

Yeah, but I think with activation checkpointing that mostly goes away. Even before fine-tuning Llama3-8B on C4 was just
67.501 GB vs 69.850 GB with activation checkpointing.
